java/src/main/java/breakfastapp/BreakfastWorkflowImpl.java

```diff
@@ -0,0 +1,50 @@
+package breakfastapp;
+
+import io.temporal.activity.ActivityOptions;
+import io.temporal.failure.ActivityFailure;
+import io.temporal.workflow.Workflow;
+import io.temporal.workflow.Saga;
+import io.temporal.common.RetryOptions;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+
+public class BreakfastWorkflowImpl implements BreakfastWorkflow {
+    private static final String BREAKFAST = "breakast";
+    // RetryOptions specify how to automatically handle retries when Activities fail.
+    private final RetryOptions retryoptions = RetryOptions.newBuilder()
+            .setInitialInterval(Duration.ofSeconds(1))
+            .setMaximumInterval(Duration.ofSeconds(100))
+            .setBackoffCoefficient(2)
+            .setMaximumAttempts(500)
+            .build();
+    private final ActivityOptions defaultActivityOptions = ActivityOptions.newBuilder()
+            .setStartToCloseTimeout(Duration.ofSeconds(5))
+            .setRetryOptions(retryoptions)
+            .build();
+    // ActivityStubs enable calls to methods as if the Activity object is local, but actually perform an RPC.
+    private final Map<String, ActivityOptions> perActivityMethodOptions = new HashMap<String, ActivityOptions>() {{
+        put(BREAKFAST, ActivityOptions.newBuilder().setHeartbeatTimeout(Duration.ofSeconds(5)).build());
+    }};
+    private final BreakfastActivity breakfastActivity = Workflow.newActivityStub(BreakfastActivity.class, defaultActivityOptions, perActivityMethodOptions);
+
+    // The transfer method is the entry point to the Workflow.
+    // Activity method executions can be orchestrated here or from within other Activity methods.
+    @Override
+    public void makeBreakfast() {
+        // You can set parallel compensations if appropriate with the Builder
+        Saga saga = new Saga(new Saga.Options.Builder().build());
```

Comment:
FWIW, this can be easily created in every language and maybe how you want to do it in each language.

Here's a couple I just hand typed. E.g. Go:

```go
type Saga []any

func (s *Saga) AddCompensation(activity any) { *s = append(*s, activity) }

func (s Saga) Compensate(ctx workflow.Context) {
  for i := len(s) - 1; i >= 0; i-- {
     // exec s[i]
  }
}
```

E.g. Python:

```python
@dataclass
class Saga:
    compensations: List[Callable] = field(default_factory=list)

    def add_compensation(self, activity: Callable) -> None:
        self.compensations.append(activity)

    async def compensate(self) -> None:
        for activity in reversed(self.compensations):
            // execute activity
```

I may have these wrong because I just hand typed, but this shows why we don't have a helper in each language. This is just too easy to implement yourself with a few lines (not to mention different people want to do different things with errors from compensations). This is kinda why I hate the buzz word saga because what we're taking it to mean here is "run some functions on error" which is a pretty basic programming concept.

Reply:
If you recall I literally have a class in the Python part that does exactly what you typed. It's called `Compensations`.  I held off on doing it in TypeScript because I was uncertain if making a class to do what was very TypeScript-y. Should I add it? I am not deep into the TypeScript world so I don't have a good sense of what "feels" right.

I recognize these are not complicated, but also, it wouldn't kill us to add it. Think about the number of times developers have to implement this themselves over and over again. Yes, it's a handful of for-loops, but if you think of the number of developers that are dealing with this problem, that adds up to hours and hours of human time. We, as an SDK provider, are tasked with providing helpful interfaces, do enable developers to solve common problems, such as Sagas. We're not paying by the "byte" to ship that little bit of extra code! 😉